internal/reconcile/install.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcile
+
+import (
+	"context"
+
+	"github.com/fluxcd/pkg/runtime/logger"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/fluxcd/pkg/runtime/conditions"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta2"
+	"github.com/fluxcd/helm-controller/internal/action"
+)
+
+type Install struct {
+	configFactory *action.ConfigFactory
+}
+
+func (r *Install) Reconcile(ctx context.Context, req *Request) error {
+	var (
+		cur    = req.Object.Status.Current.DeepCopy()
+		logBuf = action.NewLogBuffer(action.LogV(ctrl.LoggerFrom(ctx), logger.InfoLevel), 10)
+		cfg    = r.configFactory.Build(logBuf.Log, observeRelease(req.Object))
+	)
+
+	// Run install action.
+	rls, err := action.Install(ctx, cfg, req.Object, req.Chart, req.Values)
+	if err != nil {
+		// Mark failure on object.
+		req.Object.Status.Failures++
+		conditions.MarkFalse(req.Object, helmv2.ReleasedCondition, helmv2.InstallFailedReason, err.Error())
+
+		// Return error if we did not store a release, as this does not
+		// require remediation and the caller should e.g. retry.
+		if newCur := req.Object.Status.Current; newCur == nil || cur == newCur {
+			return err
+		}
+
+		// Count install failure on object, this is used to determine if
+		// we should retry the install and/or remediation.
+		req.Object.Status.InstallFailures++
```

Comment:
Do you mind clarifying why an error on `action.Install` would lead to incrementing `req.Object.Status.Failures` but not `req.Object.Status.InstallFailures` - when the release is not stored.

Reply:
The `<Install|Upgrade>Failures` are used to determine if we should stop after an `x` number of retries, and is also related to remediation. This stop is mostly present to not cause immense drift to the Helm storage (in terms of `version++`).

Given this, and the fact that  remediation is not required if there is no new object in the storage (because Helm did not apply anything, as this is a prerequisite of it applying things). Anything that does not touch the storage is better retried (indefinitely), than it to stop and wait for a human, and thus not counted as a "full attempt".

Reply:
That's great context, thank you. I think the part around the motivation for this would be good to also be captured in that comment - specially the part around retrying indefinitely.